primitives/api/proc-macro/src/decl_runtime_apis.rs

```diff
@@ -146,146 +112,74 @@ impl Fold for ReplaceBlockWithNodeBlock {
-fn fn_arg_replace_block_with_node_block(fn_arg: FnArg) -> FnArg {
-	let mut replace = ReplaceBlockWithNodeBlock {};
-	fold::fold_fn_arg(&mut replace, fn_arg)
-}
-
-/// Replace all occurrences of `Block` with `NodeBlock`
-fn return_type_replace_block_with_node_block(return_type: ReturnType) -> ReturnType {
-	let mut replace = ReplaceBlockWithNodeBlock {};
-	fold::fold_return_type(&mut replace, return_type)
-}
-
-/// Generate the functions that generate the native call closure for each trait method.
-fn generate_native_call_generators(decl: &ItemTrait) -> Result<TokenStream> {
-	let fns = decl.items.iter().filter_map(|i| match i {
-		TraitItem::Method(ref m) => Some(&m.sig),
-		_ => None,
-	});
-
-	let mut result = Vec::new();
-	let trait_ = &decl.ident;
-	let crate_ = generate_crate_access(HIDDEN_INCLUDES_ID);
-
-	// Auxiliary function that is used to convert between types that use different block types.
-	// The function expects that both are convertible by encoding the one and decoding the other.
-	result.push(quote!(
-		#[cfg(any(feature = "std", test))]
-		fn convert_between_block_types
-			<I: #crate_::Encode, R: #crate_::Decode, F: FnOnce(#crate_::codec::Error) -> #crate_::ApiError>(
-				input: &I,
-				map_error: F,
-			) -> std::result::Result<R, #crate_::ApiError>
-		{
-			<R as #crate_::DecodeLimit>::decode_with_depth_limit(
-				#crate_::MAX_EXTRINSIC_DEPTH,
-				&mut &#crate_::Encode::encode(input)[..],
-			).map_err(map_error)
+/// Versioned API traits are used to catch missing methods when implementing a specific version of a
+/// versioned API. They contain all non-versioned methods (aka stable methods) from the main trait
+/// and all versioned methods for the specific version. This means that there is one trait for each
+/// version mentioned in the trait definition. For example:
+/// ```ignore
+/// // The trait version implicitly is 1
+/// decl_runtime_apis!(
+/// 	trait SomeApi {
+/// 		fn method1(); 	// this is a 'stable method'
+///
+/// 		#[api_version(2)]
+/// 		fn method2();
+///
+/// 		#[api_version(2)]
+/// 		fn method3();
+///
+/// 		#[api_version(3)]
+/// 		fn method4();
+/// 	}
+/// );
+/// ```
+/// This trait has got three different versions. The function below will generate the following
+/// code:
+/// ```
+/// trait SomeApiV1 {
+/// 	// in V1 only the stable methods are required. The rest has got default implementations.
+/// 	fn method1();
+/// }
+///
+/// trait SomeApiV2 {
+/// 	// V2 contains all methods from V1 and V2. V3 not required so they are skipped.
+/// 	fn method1();
+/// 	fn method2();
+/// 	fn method3();
+/// }
+///
+/// trait SomeApiV3 {
+/// 	// And V3 contains all methods from the trait.
+/// 	fn method1();
+/// 	fn method2();
+/// 	fn method3();
+/// 	fn method4();
+/// }
+/// ```
+fn generate_versioned_api_traits(
+	api: ItemTrait,
+	methods: BTreeMap<u64, Vec<TraitItemMethod>>,
+) -> Vec<ItemTrait> {
+	let process_method = |mut m: TraitItemMethod| {
+		m.default = None;
```

Comment:
Why do we need this?

Reply:
This is a left over from my initial implementation. I'll remove it.